wpiutil/src/main/native/include/wpi/array.h

```diff
@@ -0,0 +1,98 @@
+// Copyright (c) FIRST and other WPILib contributors.
+// Open Source Software; you can modify and/or share it under the terms of
+// the WPILib BSD license file in the root directory of this project.
+
+#pragma once
+
+#include <array>
+#include <cstddef>
+#include <tuple>
+#include <utility>
+
+namespace wpi {
+
+/**
+ * This class is a wrapper around std::array that does compile time size
+ * checking.
+ *
+ * std::array's implicit constructor can lead result in uninitialized elements
+ * if the number of arguments doesn't match the std::array size.
+ */
+template <typename T, size_t N>
+class array : public std::array<T, N> {
+ public:
+  array() = default;
```

Comment:
Won't this let {} pass successfully?

Reply:
Yea... in that case, we need to fix the places we use the default constructor.

Reply:
I don't see a clean way of not using the default ctor in SwerveDriveKinematics.inc:41
```cpp
  wpi::array<SwerveModuleState, NumModules> moduleStates;

  for (size_t i = 0; i < NumModules; i++) {
    units::meters_per_second_t x =
        units::meters_per_second_t{moduleStatesMatrix(i * 2, 0)};
    units::meters_per_second_t y =
        units::meters_per_second_t{moduleStatesMatrix(i * 2 + 1, 0)};

    auto speed = units::math::hypot(x, y);
    Rotation2d rotation{x.to<double>(), y.to<double>()};

    moduleStates[i] = {speed, rotation};
  }
```

Reply:
How about a constructor tag to explicitly request an unfilled array?  E.g.
```
class array {
  public:
    struct empty_t {};
    constexpr empty_t empty = empty_t{};

    explicit array(empty_t) {}
```

That forces the empty case to be explicitly constructed as `array(array::empty)`

Reply:
I like this approach, but it _is_ rather verbose.
```cpp
  wpi::array<SwerveModuleState, NumModules> moduleStates{
      wpi::array<SwerveModuleState, NumModules>::empty_array};
```